src/edu/washington/escience/myria/operator/SampledDbInsertTemp.java

```diff
@@ -0,0 +1,221 @@
+   *          the key of the table that the tuples should be inserted into.
+   * @param countRelationKey
+   *          the key of the table that the tuple counts will be inserted into.
+   * @param connectionInfo
+   *          the parameters of the database connection.
+   */
+  public SampledDbInsertTemp(final Operator child, final int streamSampleSize,
+      final RelationKey sampleRelationKey, final RelationKey countRelationKey,
+      final ConnectionInfo connectionInfo) {
+    super(child);
+    // Sampling setup.
+    Preconditions.checkArgument(streamSampleSize >= 0L,
+        "sampleSize must be non-negative");
+    this.streamSampleSize = streamSampleSize;
+    doneSamplingFromChild = false;
+
+    // Relation setup.
+    Objects.requireNonNull(sampleRelationKey, "sampleRelationKey");
+    this.sampleRelationKey = sampleRelationKey;
+    Objects.requireNonNull(countRelationKey, "countRelationKey");
+    this.countRelationKey = countRelationKey;
+    this.connectionInfo = connectionInfo;
+  }
+
+  @Override
+  protected TupleBatch fetchNextReady() throws DbException {
+    if (!doneSamplingFromChild) {
+      fillReservoir();
+      batches = reservoir.getAll();
+      // Insert sampled tuples into sampleRelationKey
+      while (batchNum < batches.size()) {
+        TupleBatch batch = batches.get(batchNum);
+        accessMethod.tupleBatchInsert(sampleRelationKey, batch);
+        batchNum++;
+      }
+
+      // Write (WorkerID, PartitionSize, StreamSize) to countRelationKey
+      IntColumnBuilder wIdCol = new IntColumnBuilder();
+      IntColumnBuilder tupCountCol = new IntColumnBuilder();
+      IntColumnBuilder streamSizeCol = new IntColumnBuilder();
+      wIdCol.appendInt(getNodeID());
+      tupCountCol.appendInt(tupleCount);
+      streamSizeCol.appendInt(reservoir.numTuples());
+      ImmutableList.Builder<Column<?>> columns = ImmutableList.builder();
+      columns.add(wIdCol.build(), tupCountCol.build(), streamSizeCol.build());
+      TupleBatch tb = new TupleBatch(COUNT_SCHEMA, columns.build());
+      accessMethod.tupleBatchInsert(countRelationKey, tb);
+    }
+    return null;
+  }
+
+  /**
+   * Fills reservoir with child tuples.
+   *
+   * @throws DbException
+   *           if TupleBatch fails to get nextReady
+   */
+  private void fillReservoir() throws DbException {
+    Random rand = new Random();
+    for (TupleBatch tb = getChild().nextReady(); tb != null; tb = getChild()
+        .nextReady()) {
+      final List<? extends Column<?>> columns = tb.getDataColumns();
+      for (int i = 0; i < tb.numTuples(); i++) {
+        if (reservoir.numTuples() < streamSampleSize) {
+          // Reservoir size < k. Add this tuple.
+          for (int j = 0; j < tb.numColumns(); j++) {
+            reservoir.put(j, columns.get(j), i);
+          }
+        } else {
+          // Replace probabilistically
+          int replaceIdx = rand.nextInt(tupleCount);
+          if (replaceIdx < reservoir.numTuples()) {
```

Comment:
i think it might be clearer to use `streamSampleSize` instead of `reservoir.numTuples()`, otherwise the reader might get the impression that `reservoir.numTuples()` could be something other than `streamSampleSize` at this point in the code.